src/error.rs

```diff
@@ -1,103 +1,257 @@
-				write!(f, "Missing closing brace")
-			},
-			ErrorInner::NoSuchVariable { position: _, name } => {
-				write!(f, "No such variable: ${}", name)
-			},
+		match self {
+			Self::InvalidEscapeSequence(e) => e.fmt(f),
+			Self::MissingVariableName(e) => e.fmt(f),
+			Self::UnexpectedCharacter(e) => e.fmt(f),
+			Self::MissingClosingBrace(e) => e.fmt(f),
+			Self::NoSuchVariable(e) => e.fmt(f),
 		}
 	}
 }
 
-impl Error {
-	/// Write source highlighting for the error location.
+/// The input string contains an invalid escape sequence.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(Eq, PartialEq))]
+pub struct InvalidEscapeSequence {
+	/// The byte offset within the input where the error occurs.
 	///
-	/// The highlighting ends with a newline.
-	pub fn write_source_highlighting(&self, f: &mut impl std::fmt::Write, source: &[u8]) -> std::fmt::Result {
-		let (line, start, len) = match &self.inner {
-			ErrorInner::InvalidEscapeSequence { position, character } => {
-				let line = get_line(source, *position);
-				if character.is_some() {
-					(line, *position, 2)
+	/// This points to the associated backslash character in the source text.
+	pub position: usize,
+
+	/// The byte value of the invalid escape sequence.
+	pub character: Option<u8>,
+}
+
+impl std::error::Error for InvalidEscapeSequence {}
+
+impl std::fmt::Display for InvalidEscapeSequence {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+		if let Some(c) = self.character {
+			write!(f, "Invalid escape sequence: \\{}", char::from(c))
+		} else {
+			write!(f, "Invalid escape sequence: missing escape character")
+		}
+	}
+}
+
+/// The input string contains a variable placeholder without a variable name (`"${}"`).
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(Eq, PartialEq))]
+pub struct MissingVariableName {
+	/// The byte offset within the input where the error occurs.
+	///
+	/// This points to the `$` sign with a missing variable name in the input text.
+	pub position: usize,
+
+	/// The length of the variable placeholder in bytes.
+	pub len: usize,
+}
+
+impl std::error::Error for MissingVariableName {}
+
+impl std::fmt::Display for MissingVariableName {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+		write!(f, "Missing variable name")
+	}
+}
+
+/// The input string contains an unexpected character.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(Eq, PartialEq))]
+pub struct UnexpectedCharacter {
+	/// The byte offset within the input where the error occurs.
+	///
+	/// This points to the unexpected character in the input text.
+	pub position: usize,
+
+	/// The byte value of the unexpected character.
+	///
+	/// For multi-byte UTF-8 sequences, this only gives the value of the start byte.
+	/// You can use the `position` to get the full UTF-8 sequence from the original input string.
+	pub character: u8,
+
+	/// A human readable message about what was expected instead.
+	pub expected: ExpectedCharacter,
```

Comment:
Why not `&'static str`?

Reply:
Yeah, I was doubting here. But I wanted some flexibility to change it in the future. That's why `ExpectedCharacter` has a private field to hide the implementation details.

Reply:
As I see it, the main advantage of the new type approach is that it can also be read by the program, unlike `&str` which is only a description for humans. For example, in the future we can do:

```rust
pub enum ExpectedCharacter {
    ClosingBrace,
    Colon
}
```

In this case however, I don't see how a program would make use of this distinction — like, if we want to implement something like `cargo fix`, then yes it'd be very helpful to programmatically know what is expected, that way we can fix the error automatically. But otherwise, what *can* we really do with the information besides printing an error message anyway?

As a prior art, [`serde::de::Visitor`](https://docs.rs/serde/latest/serde/de/trait.Visitor.html) has a `expecting` method that formats a string.

But of course I could have missed other potential use cases entirely, so feel free to ignore this ;) Generally I'm all for future-proofing anyway.

Reply:
For example, I was thinking of potentially also providing a `expected_chars() -> &[char]` in the future.